src/librustc_typeck/collect.rs

```diff
@@ -546,12 +546,14 @@ fn convert_struct_variant<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
             vis: ty::Visibility::from_hir(&f.vis, node_id, tcx)
         }
     }).collect();
+
     ty::VariantDef {
         did,
         name,
         discr,
         fields,
         ctor_kind: CtorKind::from_hir(def),
+        can_extend_field_list: tcx.has_attr(did, "non_exhaustive"),
```

Comment:
One weird thing -- I'm not sure how this will work -- the `did` in the case of a tuple-struct is I think the synthetic def-id creator for the struct constructor? We just need to be sure to test `#[non_exhaustive] pub struct Foo(pub u32)`

Reply:
well I guess we have some tests for that already so... seems ok.